File: 0216-combination-sum-iii/0216-combination-sum-iii.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** combinationSum3(int k, int n, int* returnSize, int** returnColumnSizes) {
    int maxCombinations = 100;  // Initial capacity
    int** result = (int**)malloc(maxCombinations * sizeof(int*));
    *returnColumnSizes = (int*)malloc(maxCombinations * sizeof(int));
    *returnSize = 0;
    
    // Current combination being built
    int* current = (int*)malloc(k * sizeof(int));
    
    void backtrack(int start, int remainSum, int pos) {
        // Base cases
        if (remainSum < 0 || pos > k) return;
        if (pos == k) {
            if (remainSum == 0) {
                // Valid combination found
                result[*returnSize] = (int*)malloc(k * sizeof(int));
                memcpy(result[*returnSize], current, k * sizeof(int));
                (*returnColumnSizes)[*returnSize] = k;
                (*returnSize)++;
            }
            return;
        }
        
        // Try each possible number
        for (int i = start; i <= 9; i++) {
            current[pos] = i;
            backtrack(i + 1, remainSum - i, pos + 1);
        }
    }
    
    backtrack(1, n, 0);
    
    free(current);
    return result;
}

// Test function
void runTest(int k, int n) {
    int returnSize;
    int* returnColumnSizes;
    int** result = combinationSum3(k, n, &returnSize, &returnColumnSizes);
    
    printf("k = %d, n = %d\n", k, n);
    printf("Output:\n");
    for (int i = 0; i < returnSize; i++) {
        printf("[");
        for (int j = 0; j < returnColumnSizes[i]; j++) {
            printf("%d", result[i][j]);
            if (j < returnColumnSizes[i] - 1) printf(",");
        }
        printf("]\n");
        free(result[i]);
    }
    printf("\n");
    
    free(result);
    free(returnColumnSizes);
}

File: 0216-combination-sum-iii/0216-combination-sum-iii.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(k, n, 1, new ArrayList<>(), result);
        return result;
    }
    
    private void backtrack(int k, int remainSum, int start, 
                          List<Integer> current, 
                          List<List<Integer>> result) {
        // Base cases
        if (remainSum < 0 || current.size() > k) return;
        
        if (current.size() == k) {
            if (remainSum == 0) {
                result.add(new ArrayList<>(current));
            }
            return;
        }
        
        // Try each possible number
        for (int i = start; i <= 9; i++) {
            current.add(i);
            backtrack(k, remainSum - i, i + 1, current, result);
            current.remove(current.size() - 1);  // backtrack
        }
    }
}